deflate.c

```diff
@@ -1246,22 +1246,35 @@ void check_match(deflate_state *s, IPos start, IPos match, int length) {
  *    option -- not supported here).
  */
 
-void ZLIB_INTERNAL fill_window_c(deflate_state *s) {
+void ZLIB_INTERNAL fill_window(deflate_state *s) {
     unsigned n;
-    unsigned more;    /* Amount of free space at the end of the window. */
+    unsigned long more;    /* Amount of free space at the end of the window. */
     unsigned int wsize = s->w_size;
 
     Assert(s->lookahead < MIN_LOOKAHEAD, "already enough lookahead");
 
     do {
-        more = (unsigned)(s->window_size -(unsigned long)s->lookahead -(unsigned long)s->strstart);
+        more = s->window_size - s->lookahead - s->strstart;
+
+        /* Deal with !@#$% 64K limit: */
+        if (sizeof(int) <= 2) {
+            if (more == 0 && s->strstart == 0 && s->lookahead == 0) {
+                more = wsize;
+
+            } else if (more == (unsigned)(-1)) {
+                /* Very unlikely, but possible on 16 bit machine if
+                 * strstart == 0 && lookahead == 1 (input done a byte at time)
+                 */
+                more--;
+            }
+        }
```

Comment:
This 64K limit block can be removed entirely. sizeof(int) is never smaller than 4 on the platforms we support.

Reply:
Ok I wasn't sure because I noticed it in some of _fill_window_ implementations but not in others.

Reply:
But should probably be removed from its source as a separate commit since it is a separate logical change.

Reply:
We actually already removed it, but we missed one and never got around to remove it :)

Reply:
Ah that makes sense. Do you still want it removed in a separate commit?

Reply:
Well, since you rewrite it completely and base it on fill_window() that already had removed it and deleted the file where it was, I guess it is not needed.
I hadn't completed the review yet, but now I have and have all the context.

Reply:
Yeah, I tried to merge the best code from the various implementations of _fill_window_, that is why some of the changes are in _fill_window_c_ and wasn't sure about this additional check.

Reply:
I have updated the PR.